### PR TITLE
Initialize hostile mobs

### PR DESCRIPTION
**Pull request by moxian**
## What Does This PR Do
`/mob/living/simple_animal/hostile` and its children are `/Initialize` now.
.. except for `/mob/living/simple_animal/hostile/spaceWorm` because.. reasons

Lightly tested; seems to work fine.
Action buttons definitely work, and GC is reasonably happy.

## Why It's Good For The Game
Good practice

## Changelog
N/A

### CODE REVIEW

File: code/modules/mining/minebot.dm [hunk 1]
```diff
@@ -40,22 +40,18 @@
...
 	stored_gun = new(src)
-	toggle_light_action = new()
-	toggle_light_action.Grant(src)
-	toggle_meson_vision_action = new()
-	toggle_meson_vision_action.Grant(src)
-	toggle_mode_action = new()
-	toggle_mode_action.Grant(src)
-	dump_ore_action = new()
-	dump_ore_action.Grant(src)
+	for(var/action_path in list(
```
Comment (farie82) [inline code/modules/mining/minebot.dm:47]: Can you make this into a static list in this proc? Should make it a bit more readable

File: code/modules/mob/living/simple_animal/hostile/bees.dm [hunk 2]
```diff
@@ -59,11 +59,14 @@
 /mob/living/simple_animal/hostile/poison/bees/Process_Spacemove(movement_dir = 0)
 	return TRUE
 
-/mob/living/simple_animal/hostile/poison/bees/New()
-	..()
-	generate_bee_visuals()
+/mob/living/simple_animal/hostile/poison/bees/ComponentInitialize()
+	. = ..()
```
Comment (farie82) [inline code/modules/mob/living/simple_animal/hostile/bees.dm:63]:
```suggestion
	..()
```

File: code/modules/mob/living/simple_animal/hostile/hellhound.dm [hunk 3]
```diff
@@ -33,12 +33,11 @@
...
 	var/smoke_lastuse = 0
 	var/smoke_freq = 300 // 30 seconds
-	var/datum/action/innate/demon/whisper/whisper_action
 	footstep_type = FOOTSTEP_MOB_CLAW
 
-/mob/living/simple_animal/hostile/hellhound/New()
+/mob/living/simple_animal/hostile/hellhound/Initialize(mapload)
 	. = ..()
-	whisper_action = new()
+	var/datum/action/whisper_action = new /datum/action/innate/demon/whisper()
```
Comment (farie82) [inline code/modules/mob/living/simple_animal/hostile/hellhound.dm:40]:
```suggestion
	var/datum/action/innate/demon/whisper/whisper_action = new
```

File: code/modules/mining/minebot.dm [hunk 4]
```diff
@@ -40,22 +40,18 @@
...
+		/datum/action/innate/minedrone/toggle_light,
+		/datum/action/innate/minedrone/toggle_meson_vision,
+		/datum/action/innate/minedrone/toggle_mode,
+		/datum/action/innate/minedrone/dump_ore,
+	))
+		var/datum/action/act = new action_path()
```
Comment (farie82) [inline code/modules/mining/minebot.dm:53]:
```suggestion
		var/datum/action/act = new action_path
```

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/brown.dm [hunk 5]
```diff
@@ -27,9 +27,8 @@
...
 	web_type = null
-	var/datum/action/innate/terrorspider/ventsmash/ventsmash_action
 
-/mob/living/simple_animal/hostile/poison/terror_spider/brown/New()
-	..()
-	ventsmash_action = new()
-	ventsmash_action.Grant(src)
+/mob/living/simple_animal/hostile/poison/terror_spider/brown/Initialize(mapload)
+	. = ..()
+	var/datum/action/act = new /datum/action/innate/terrorspider/ventsmash()
```
Comment (farie82) [inline code/modules/mob/living/simple_animal/hostile/terror_spiders/brown.dm:33]:
```suggestion
	var/datum/action/innate/terrorspider/ventsmash/act = new
```

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/empress.dm [hunk 6]
```diff
@@ -28,16 +28,14 @@
...
 
-/mob/living/simple_animal/hostile/poison/terror_spider/queen/empress/New()
-	..()
-	empresslings_action = new()
-	empresslings_action.Grant(src)
-	empresserase_action = new()
-	empresserase_action.Grant(src)
+/mob/living/simple_animal/hostile/poison/terror_spider/queen/empress/Initialize(mapload)
+	. = ..()
+	var/datum/action/act_ling = new /datum/action/innate/terrorspider/queen/empress/empresslings()
```
Comment (farie82) [inline code/modules/mob/living/simple_animal/hostile/terror_spiders/empress.dm:35]:
```suggestion
	var/datum/action/innate/terrorspider/queen/empress/empresslings/act_ling = new
```
Also wow that type. `empresslings`

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/empress.dm [hunk 7]
```diff
@@ -28,16 +28,14 @@
...
+	act_ling.Grant(src)
+	var/datum/action/act_erase = new /datum/action/innate/terrorspider/queen/empress/empresserase()
```
Comment (farie82) [inline code/modules/mob/living/simple_animal/hostile/terror_spiders/empress.dm:37]:
```suggestion
	var/datum/action/innate/terrorspider/queen/empress/empresserase/act_erase = new
```

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/green.dm [hunk 8]
```diff
@@ -22,13 +22,12 @@
...
-	var/datum/action/innate/terrorspider/greeneggs/greeneggs_action
 
 
-/mob/living/simple_animal/hostile/poison/terror_spider/green/New()
-	..()
-	greeneggs_action = new()
-	greeneggs_action.Grant(src)
+/mob/living/simple_animal/hostile/poison/terror_spider/green/Initialize(mapload)
+	. = ..()
+	var/datum/action/act = new /datum/action/innate/terrorspider/greeneggs()
```
Comment (farie82) [inline code/modules/mob/living/simple_animal/hostile/terror_spiders/green.dm:29]:
```suggestion
	var/datum/action/innate/terrorspider/greeneggs/act = new
```

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/mother.dm [hunk 9]
```diff
@@ -28,26 +28,21 @@
...
+/mob/living/simple_animal/hostile/poison/terror_spider/mother/Initialize(mapload)
+	. = ..()
+	for(var/action_path in list(
+		/datum/action/innate/terrorspider/ventsmash,
+		/datum/action/innate/terrorspider/remoteview,
+		/datum/action/innate/terrorspider/mother/royaljelly,
+		/datum/action/innate/terrorspider/mother/gatherspiderlings,
+		/datum/action/innate/terrorspider/mother/incubateeggs,
+	))
+		var/datum/action/act = new action_path()
```
Comment (farie82) [inline code/modules/mob/living/simple_animal/hostile/terror_spiders/mother.dm:44]:
```suggestion
		var/datum/action/act = new action_path
```

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/mother.dm [hunk 10]
```diff
@@ -28,26 +28,21 @@
...
-	remoteview_action.Grant(src)
-	royaljelly_action = new()
-	royaljelly_action.Grant(src)
-	gatherspiderlings_action = new()
-	gatherspiderlings_action.Grant(src)
-	incubateeggs_action = new()
-	incubateeggs_action.Grant(src)
+/mob/living/simple_animal/hostile/poison/terror_spider/mother/Initialize(mapload)
+	. = ..()
+	for(var/action_path in list(
```
Comment (farie82) [inline code/modules/mob/living/simple_animal/hostile/terror_spiders/mother.dm:37]: Here as well. Can you make this a static list?

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/queen.dm [hunk 11]
```diff
@@ -50,15 +50,13 @@
...
-	var/datum/action/innate/terrorspider/ventsmash/ventsmash_action
-	var/datum/action/innate/terrorspider/remoteview/remoteview_action
 
 
-/mob/living/simple_animal/hostile/poison/terror_spider/queen/New()
-	..()
-	ventsmash_action = new()
+/mob/living/simple_animal/hostile/poison/terror_spider/queen/Initialize(mapload)
+	. = ..()
+	var/datum/action/ventsmash_action = new /datum/action/innate/terrorspider/ventsmash/()
```
Comment (farie82) [inline code/modules/mob/living/simple_animal/hostile/terror_spiders/queen.dm:57]:
```suggestion
	var/datum/action/innate/terrorspider/ventsmash/ventsmash_action = new
```

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/queen.dm [hunk 12]
```diff
@@ -50,15 +50,13 @@
...
 	ventsmash_action.Grant(src)
-	remoteview_action = new()
+	var/datum/action/remoteview_action = new /datum/action/innate/terrorspider/remoteview()
```
Comment (farie82) [inline code/modules/mob/living/simple_animal/hostile/terror_spiders/queen.dm:59]:
```suggestion
	var/datum/action/innate/terrorspider/remoteview/remoteview_action = new
```

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/terror_spiders.dm [hunk 13]
```diff
@@ -262,21 +260,21 @@ GLOBAL_LIST_EMPTY(ts_spiderling_list)
...
 	GLOB.ts_spiderlist += src
 	add_language("Spider Hivemind")
 	if(spider_tier >= TS_TIER_2)
 		add_language("Galactic Common")
 	default_language = GLOB.all_languages["Spider Hivemind"]
 
 	if(web_type)
-		web_action = new()
-		web_action.Grant(src)
+		var/datum/action/web_act = new /datum/action/innate/terrorspider/web()
```
Comment (farie82) [inline code/modules/mob/living/simple_animal/hostile/terror_spiders/terror_spiders.dm:272]:
```suggestion
		var/datum/action/innate/terrorspider/web/web_act = new
```

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/terror_spiders.dm [hunk 14]
```diff
@@ -262,21 +260,21 @@ GLOBAL_LIST_EMPTY(ts_spiderling_list)
...
+		web_act.Grant(src)
 	if(regen_points_per_tick < regen_points_per_hp)
 		// Only grant the Wrap action button to spiders who need to use it to regenerate their health
-		wrap_action = new()
-		wrap_action.Grant(src)
+		var/datum/action/wrap_act = new /datum/action/innate/terrorspider/wrap()
```
Comment (farie82) [inline code/modules/mob/living/simple_animal/hostile/terror_spiders/terror_spiders.dm:276]:
```suggestion
		var/datum/action/innate/terrorspider/wrap/wrap_act = new
```

**Review by farie82 — CHANGES_REQUESTED** [review]
Can you confirm that your GC-related changes fix an existing GC issue?

**Review by moxian — COMMENTED** [review]
Most of them don't fix an existing GC issue, as actions are cleaned up during `/mob/Destroy()`.

However it feels like a good GC practice to do *something* with all nontrivial vars on `/Destroy()`, just so that future code readers don't need to spend time figuring out that there's no bug.
There's no need to hold on to the vars, since no functionality of the mobs use them beyond initial addition, which is why i removed those.

The guardian `beacon` turf ref *might* hypothetically cause an issue whenever the turf is deleted, but i'm not well enough versed in turf code to confirm. And regardless, even if it does happen, it should be extremely rare.

The statue's `.creator` *should* be a real issue, however the steps are:
- flesh-to-stone someone
- staff of animation said someone
- gib self
- have the round not immediately end

which is rather contrived. Besides, something's funky is going on with projectile code, so the `.firer` var of the animation bolt (which later becomes statue's `.creator`) is somehow set to `null` so i can't 100% confirm in test server that this is a problem.

TL;DR: those are not "real" issues in the sense that you are extremely unlikely to observe them.

**Comment by moxian** [comment]
On master, spawning then deleting a brown spider:

<details>
```
[2022-07-24T09:43:49] Starting up. Round ID is NULL
[2022-07-24T09:43:49] -------------------------
[2022-07-24T09:47:20] Beginning search for references to a /datum/mind.
[2022-07-24T09:47:20] Finished searching globals
[2022-07-24T09:47:20] Found /datum/mind [0x21008385] in /mob/dead/observer's [0x300000a] mind var. World -> /mob/dead/observer
[2022-07-24T09:48:02] Finished searching atoms
[2022-07-24T09:48:04] Finished searching datums
[2022-07-24T09:48:04] Finished searching clients
[2022-07-24T09:48:04] Completed search for references to a /datum/mind.
[2022-07-24T09:48:04] GC: -- [0x2100000d] | /datum/mind was unable to be GC'd --
[2022-07-24T09:56:40] Beginning search for references to a /datum/action/innate/terrorspider/wrap.
[2022-07-24T09:56:40] Finished searching globals
[2022-07-24T09:56:40] Found /datum/action/innate/terrorspider/wrap [0x21009908] in /mob/living/simple_animal/hostile/poison/terror_spider/brown's [0x3000009] wrap_action var. World -> /mob/living/simple_animal/hostile/poison/terror_spider/brown
[2022-07-24T09:57:24] Finished searching atoms
[2022-07-24T09:57:25] Finished searching datums
[2022-07-24T09:57:25] Finished searching clients
[2022-07-24T09:57:25] Completed search for references to a /datum/action/innate/terrorspider/wrap.
[2022-07-24T09:57:25] GC: -- [0x2100000d] | /datum/action/innate/terrorspider/wrap was unable to be GC'd --
[2022-07-24T09:57:41] Beginning search for references to a /datum/action/innate/terrorspider/ventsmash.
[2022-07-24T09:57:41] Finished searching globals
[2022-07-24T09:57:41] Found /datum/action/innate/terrorspider/ventsmash [0x210083d6] in /mob/living/simple_animal/hostile/poison/terror_spider/brown's [0x3000009] ventsmash_action var. World -> /mob/living/simple_animal/hostile/poison/terror_spider/brown
[2022-07-24T09:58:28] Finished searching atoms
[2022-07-24T09:58:30] Finished searching datums
[2022-07-24T09:58:30] Finished searching clients
[2022-07-24T09:58:30] Completed search for references to a /datum/action/innate/terrorspider/ventsmash.
[2022-07-24T09:58:30] GC: -- [0x2100000d] | /datum/action/innate/terrorspider/ventsmash was unable to be GC'd --
```
</details>

With the PR:

<details>
```
[2022-07-24T10:08:56] Starting up. Round ID is NULL
[2022-07-24T10:08:56] -------------------------
[2022-07-24T10:18:12] Beginning search for references to a /datum/mind.
[2022-07-24T10:18:12] Finished searching globals
[2022-07-24T10:18:12] Found /datum/mind [0x21005c7a] in /mob/dead/observer's [0x3000007] mind var. World -> /mob/dead/observer
[2022-07-24T10:18:58] Finished searching atoms
[2022-07-24T10:19:00] Finished searching datums
[2022-07-24T10:19:00] Finished searching clients
[2022-07-24T10:19:00] Completed search for references to a /datum/mind.
[2022-07-24T10:19:00] GC: -- [0x2100000d] | /datum/mind was unable to be GC'd --
```
</details>

Here's the PR with spawning tspider, VV'ing it and *then* deleting:
<details>
```
[2022-07-24T10:33:54] Beginning search for references to a /mob/living/simple_animal/hostile/poison/terror_spider/brown.
[2022-07-24T10:33:54] Finished searching globals
[2022-07-24T10:34:31] Finished searching atoms
[2022-07-24T10:34:32] Finished searching datums
[2022-07-24T10:34:32] Finished searching clients
[2022-07-24T10:34:32] Completed search for references to a /mob/living/simple_animal/hostile/poison/terror_spider/brown.
[2022-07-24T10:34:32] GC: -- [0x2100000d] | /mob/living/simple_animal/hostile/poison/terror_spider/brown was unable to be GC'd --
```
</details>
still no actions mentions.

However.. i find this GC ref tracking tool to be somewhat unreliable. On master i spawned (and deleted) *two* brown tspiders and a mining drone but only got gc failures about one.

Anyhow...

**Comment by farie82** [comment]
> However.. i find this GC ref tracking tool to be somewhat unreliable. On master i spawned (and deleted) _two_ brown tspiders and a mining drone but only got gc failures about one.

IIRC, this might be due to how the MC handles stuff. If a SS hangs for a long time then it gets sort of punished. So it doesn't run as often. Since the reference searching happens in SSgarbage, this means that it might not run as often. Thus you'd not get the same results in the same timeframe. Running `fire` manually on SSgarbage should fix this.

But nice job with testing and thanks for confirming your expectations

**Review by farie82 — APPROVED** [review]
